### ISSUE TITLE
Recommended technique for Math formulas in EPUB

### ISSUE THREAD

**Issue by rdeltour** [w3c/epub-specs#845]
Recent discussions ([EPUB a11y call on August 24](https://docs.google.com/document/d/1GcwrWIXSTfQYJlufbceK8utU8KPdOZp2wtj2T23P9hQ/edit?pli=1#bookmark=kix.nbeymcpyngb) + [follow-up thread on the mailing list](https://groups.google.com/d/msg/epub-working-group/k8TgNoDtls4/8M_i_BvUBgAJ)) have revived the longtime question of how best to represent math formulas in EPUB. This issue intends to track the various discussions and solution candidates.
## Background

The issue of math formulas in EPUB consists of at least two distinct (but related) challenges:
1. how to visually render the formula
2. how to make it usable through assistive technologies

Historically, EPUB has promoted MathML as the preferred way to represent math. However, MathML is poorly supported by numerous Reading Systems. Many of them do not even parse the MathML to render the fallback image defined in MathML's `altimg` attribute. Many content publishers –even the ones that can produce MathML– end up commenting out the MathML and just use images of formulas instead.

It must also be noted that MathML is considered by some experts as a failing appraoch to represent math on the web (see notably @pkra's [thoughts on the topic](https://www.peterkrautzberger.org/0186/)).

As for accessibility, the situation is mixed. For some very simple formulas, simple alternative text is probably enough. For more complex content, MathML is reasonably well supported by major screen readers (sometimes with 3d party libraries like MathPlayer), despite being limited by the inherent lack of semantics in presentation MathML. Some specialized JS-based renderers (like [MathJax's accessibility extensions](https://www.mathjax.org/mathjax-accessibility-extensions-v1-now-available/)) may also be relevant alternatives.

All in all, there is no ideal solution. What is needed is one (or several?) best practices that can be recommended to content publishers in the short-to-mid term.
## Candidate techniques

The constraints are:
- the formula must be visually rendered by both legacy and newer Reading Systems
- legacy reading systems cannot be updated
- the solution is accessible (even if MathML is not supported / not available)
- authors must be able to reasonably easily create content from a MathML representation (most publishers do have MathML content).
### ~~Option 1 - `data-mathml` atttribute~~

The formula is represented with an `img` element, with alternative text in its `alt` attribute. Additionally, the equivalent MathML markup is available (escaped) in a `data-math` attribute that can be used by a javascript renderer.

Example: 

``` html
  <img alt="my equation" src="my-formula.jpg" data-math="…escaped mathml…"/>
```

Pros:
- works on legacy RS
- requires no specification work

Cons:
- usage of the `data-math` attribute is a non-standard used of the HTML spec if it is intended for use by the RS (it's probably OK if the javascript is provided by the Author).
### Option 2a - `object` element with MathML fallback

The formula is represented as an image in the `object` element, for which an alternative text can be defined with the `aria-label` attribute. The `object` can have MathML child content, which could then be used as a fallback by the RS or AT that prefers it.

Example:

``` html
<object data="image.jpg" aria-label="my equation description">
 <mathml>
   (...)
 </mathml>
</object>
```

Pros:
- works on legacy RS
- easy to author

Cons:
- requires some slightly bended interpretation of HTML, which says that the `object`'s fallback content can be shown only "if the user has indicated a preference" (for instance if "that content uses a format that the user finds more accessible").
### ~~Option 2b - `object` element with image fallback~~

The formula is represented as an external MathML document embedded with an `object` element, for which an alternative text can be defined with the `aria-label` attribute. The `object` can have an image as its child content, which could then be used as a fallback by the RS.

Example:

``` html
<object data="mathml-formula.xml" aria-label="my equation description" type="application/mathml+xml">
  <img src="image-formula.jpg" alt="my equation description"/>
</object>
```

Pros:
- works on legacy RS and current browsers

Cons:
- authoring inconvenience
### Option 3 - Media Query

 A new Media Query "detecting" MathML support. We could then assign `display:`none' to the math element and a non-none display to an image fallback sibling for legacy reading systems and have that new MQ assign a non-none display value to the math element and `display: none` to the image fallback sibling if MathML is implemented.

 See @therealglazou's [proposal](https://groups.google.com/d/msg/epub-working-group/jaSz8Fh9j90/StPZ1pPlCAAJ).

Example:

``` html
<img class="formula-image" src="formula.jpg" alt="formula description"/>
<mathml class="formula-mathml">
   …
</mathml>
```

``` css
.formula-mathml {
  display: none;
}
@media (mathml) {
  .formula-mathml {
    display: inline;
  }
  .formula-image {
    display: none;
  }
} 
```

Pros:
- conceptually simple
- similar MQs will likely be used for other feature detection (e.g. javascript)

Cons:
- requires spec work in CSS
- slightly more tedious to author
### ~~Option 4 - `aria-details`~~

The formula is rendered with an image, and the MathML is referenced from an `aria-details` attribute on the image.

Example:

``` html
<img class="formula-image" src="formula.jpg" alt="formula description" 
          aria-details="formula-details" />
<details id="formula-details">
  <summary>Example</summary>
  <mathml>
   …
  </mathml>
</details>
```
### Option 5 - custom classes in package metadata

Special CSS classes for both MathML and non-MathML formulas are declared in Package metadata. Reading systems that support MathML would override the display property to render the MathML instead of the image.

Example:
Content Doc:

``` html
<img class="math-image" src="..."/>
<span class="math-mathml">
  <m:math>...</m:math>
</span>
```

CSS:

``` css
.mathml { display: none; }
```

Package:

``` xml
<meta property="math:mathml">math-mathml</meta>
<meta property="math:no-mathml">math-image</meta>
```

A MathML-aware RS would  override the CSS based on the values defined in Metadata above:

``` css
.math-mathml { display: inline !important; }
.math-image { display: none !important; }
```
## Other references
- [EPUB Accessibility Technique DESC-002 "Include descriptions for MathML content"](http://www.idpf.org/epub/a11y/techniques/techniques.html#sec-desc-002)
- Related older issue #681 "MathML alternative content"
- ARIA Best Practices issue "Math example unclear" w3c/aria-practices#26
- [Math on WebPages meeting minutes on Aug 29](https://lists.w3.org/Archives/Public/public-mathonwebpages/2016Aug/0008.html)

**Comment by danielweck** [comment]
Additional offscreen design proposal: https://rawgit.com/mhakkinen/extdesc/master/Extended-desc.md.html

**Comment by danielweck** [comment]
MathJax's Peter K. 's thoughts on MathML Media Query:
https://www.peterkrautzberger.org/0190/
@pkra

**Comment by danielweck** [comment]
An old issue...but perhaps a good time to mention the more recent DAISY proposals:
https://github.com/daisy/epub-accessibility-tests/blob/master/content/epub30-test-0330/EPUB/Text/nav.xhtml
Test EPUB:
https://github.com/daisy/epub-accessibility-tests/blob/master/build/Advanced-Accessibility-Tests-Mathematics-v1.1.0.epub

**Comment by VeryGoodErotica** [comment]
What about this? It's based on 2a and the `role="math"` could be used as a trigger to let ePub viewers know that if they properly render MathML then do not use the image, use the MathML content. Obviously readers not updated will use the image even if they support MathML but that is not so bad.

    <object data="equation.{svg,png}" role="math" aria-label="y equals b INVISIBLE TIMES x to-the-power-of k">
      <math xmlns="http://www.w3.org/1998/Math/MathML" display="inline">
      <mi>y</mi>
      <mo>=</mo>
      <mi>b</mi>
      <mo>&#x2062;</mo>
      <msup>
        <mi>x</mi>
        <mi>k</mi>
      </msup>
      </math>
    </object>

**Comment by VeryGoodErotica** [comment]
Tested, it works in both Calibre and on my Kobo. ePubCheck complains that `role=math` is invalid for object so maybe next version of ePub should have a `epub:type="math"` that is used to let viewers know to render the MathML instead if they support it? with the MathSpeak in a SVG `desc` node like MathJax does?

**Comment by VeryGoodErotica** [comment]
Guess what I'm suggesting is option 2a but add an `epub:type="math"` attribute in next ePub version that viewers can optionally use to trigger a preference of the MathML if they properly support it.

Screen readers can be triggered by an `aria-labelledby` in the svg node,

**Comment by VeryGoodErotica** [comment]
sorry I haven't finished my coffee.

Option 2a with the `epub:type="math"` and an `aria-label="MathSpeak string"` in the object node.

**Comment by GeorgeKerscher** [comment]
Hello,

 

Attached is a test epub with the various options we are testing for screen reader users, low vision users, persons who are dyslexic, and for mainstream folks. We want to come up with a best practice.

 

I think we have one more test to add before we start our extensive testing.

 

I am liking the test 020, but we have some issues with some systems. We are trying to track that down.

 

Best

George

**Comment by laudrain** [comment]
Dear George,
Sorry, I don’t see any attachment ?
Luc

De : George <notifications@github.com>
Répondre à : w3c/publ-epub-revision <reply@reply.github.com>
Date : vendredi 27 mars 2020 à 16:51
À : w3c/publ-epub-revision <publ-epub-revision@noreply.github.com>
Cc : Subscribed <subscribed@noreply.github.com>
Objet : Re: [w3c/publ-epub-revision] Recommended technique for Math formulas in EPUB (#845)

Hello,



Attached is a test epub with the various options we are testing for screen reader users, low vision users, persons who are dyslexic, and for mainstream folks. We want to come up with a best practice.



I think we have one more test to add before we start our extensive testing.



I am liking the test 020, but we have some issues with some systems. We are trying to track that down.



Best

George

**Comment by GeorgeKerscher** [comment]
Hello,

 

If the mail server stripped out the attached EPUB, we have the math test book at:

http://www.epubtest.org/test-books

 

The advanced test book is in process of being revised, but this is a start. We hope to have the revised version available soon.

 

Best

George

**Comment by VeryGoodErotica** [comment]
Thank you very much for those working on this.

The solution I came up is currently working in the ePub viewers I personally have access to, but I _definitely_ would prefer a method with *proper* accessibility testing behind it.

**Comment by mattgarrish** [comment]
Until we have a technique that reasonably works everywhere, I don't think there's anything we can do with this issue. I'm going to put a deferred label on it as I don't see it being resolved in our current charter timeline. We can return to it whenever there's more clarity (assuming mathml support doesn't overtake the need for this).

**Comment by OriKuttner** [comment]
There used to be a discussion regarding MathML.
I think most reading applications does not support it.

-- 
Ori Idan CEO Helicon Books
http://www.heliconbooks.com

**Comment by avneeshsingh** [comment]
We developed Math techniques and have done extensive testing on it. Which will improve further with Chrome's support for MathML.
The EPUB with the examples of best practices for Math is at:
http://epubtest.org/books/Accessibility-Tests-Mathematics-v1.1.1.epub

And explanation of the techniques is at: 
http://diagramcenter.org/best-practices-for-authoring-math.html

So, we can consider closing this issue.

**Comment by alexshpilkin** [comment]
@avneeshsingh It appears that your report is, to a large extent, focused on _accessible_ maths, which is undeniably important, but also not nearly the whole problem.

And... Do I understand correctly that the proposed solution for reasonably supported maths in EPUB books is... PNG images with no possibility of progressive enhancement? (OK, PNG images with `<details>`, in some cases.) That is, something the editors of the only decently typeset ebook with non-trivial mathematics that I’ve seen (The Feynman Lectures on Physics) _explicitly rejected_ as unbearably ugly. And the alternative requires MathML support in the reader on pain of making the book impossible to understand.

I can understand that as a faithful representation of the current state of the reader world, but it’s still just so sad.

**Comment by iherman** [comment]
> I can understand that as a faithful representation of the current state of the reader world, but it’s still just so sad.

It is. But there is some light in the tunnel. W3C is discussing setting up a new MathML Working Group whose main goal is to consolidate a profile of MathML that can, and will be, available in browsers. If that happens, then the PNG fallback become less prominent.

Fingers crossed :-)

**Comment by alexshpilkin** [comment]
@iherman wrote:
>> I can understand that as a faithful representation of the current state of the reader world, but it’s still just so sad.
>
> It is. But there is some light in the tunnel. W3C is discussing setting up a new MathML Working Group whose main goal is to consolidate a profile of MathML that can, and will be, available in browsers. If that happens, then the PNG fallback become less prominent.

I’m not even particularly against a PNG _fallback_; I mean, having several representations of the same formula isn’t pretty and does add some bloat, but that’s what tooling is for. However, the discussion above leads me to believe that having a reasonably compatible, no-scripting-required MathML to SVG to PNG fallback isn’t possible with current readers, not with `<svg:switch>`, not with `<epub:switch>`, not with `<picture>`, not with `<object>`, just not possible at all in a single EPUB. If you want compatibility, it’s bare PNGs or bust. (The web is easier in that you can just code your fallback logic in JavaScript and it’ll usually work.)

Preferably such a fallback would allow for yet another alternative representation targeted at screen readers (although, TBH, a significant portion of maths that I care about best exists in a two-dimensional form—commutative diagrams, deduction trees, even similar algebraic expressions with related terms lined up—and I’ve never seen coherent guidelines as to how deal with these in an accessible way). Ideally it would even allow an additional, separate representation for copying (AsciiMath or TeX code, probably). But so far even the non-accessible fallback described above seems impossible; I’ve subscribed to this thread to see myself proven wrong on this count and haven’t succeeded so far. :(

**Comment by mattgarrish** [comment]
Closing this issue, but feel free to continue to discuss. The issue here is generally poor support for MathML and we can't solve that in the techniques at this time.

**Comment by jenstroeger** [comment]
Is there any status update on this? There’s a discussion [Best Practices for Authoring MathML in EPUB](https://daisy.github.io/transitiontoepub/best-practices/mathML/mathMLBestPractices.html) from DAISY but I can’t find a definite guideline or recommendation?

**Comment by avneeshsingh** [comment]
@jenstroeger In addition to the best practices document that you mentioned above, DAISY knowledge base also has [a section for MathML.](https://kb.daisy.org/publishing/docs/html/mathml.html)
We will be revising the mentioned best practices sometime this year to reflect latest support of browsers and assistive technologies.

**Comment by jenstroeger** [comment]
Thank you @avneeshsingh — so what’s the recommendation today to target current and likely dated reading systems?

The equation in my particular case is a simple inlined _E=mc<sup>2</sup>_ which I could just note down/mark up as
```html
<span role="math">E = mc<sup>2</sup></span>
```
to ensure that current reading systems can reliably handle it?

**Comment by avneeshsingh** [comment]
It would be interesting to know that more than 50% of the reading systems that we tested support MathML.

This change was visible after Chrome and Edge started supporting MathML natively. Because many reading systems are built on these browsers, MathML support started to grow.

So, I think that it would be OK if you keep the equation in MathML 4.

**Comment by jenstroeger** [comment]
> It would be interesting to know that more than 50% of the reading systems that we tested support MathML.

Do you know how those “more than 50% tested” correlate to the devices in use out in the wild? What if 80% (just making up a number) of the devices in use make up the other 50%? 🤔

**Comment by avneeshsingh** [comment]
In coming months, we will be posting accessibility testing results of popular reading systems on https://epubtest.org/results

It will provide more accurate information.

**Comment by GeorgeKerscher** [comment]
Hi,
 
We will know more, but I suspect that of the Reading Systems used in Higher Education that we may get close to 100%.
 
Best
George

**Comment by gregoriopellegrino** [comment]
> The equation in my particular case is a simple inlined _E=mc2_ which I could just note down/mark up as
> 
> ```
> <span role="math">E = mc<sup>2</sup></span>
> ```

Hi @jenstroeger, we have discussed this possible approach (that looked promising) in both the DAISY TIES group and W3C. Using `role="math"` to mark formulas in plain text does not seem like a good solution. The direction we give now, even for inline formulas, is to transform them to MathML.

**Comment by GeorgeKerscher** [comment]
Hi,
 
I just tested with NVDA and the span was not read. It is because of the role of math.
 
Best
George
